src/Data/Swagger/Internal.hs

```diff
@@ -1150,10 +1153,11 @@ instance ToJSON SecurityDefinitions where
   toJSON (SecurityDefinitions sd) = toJSON sd
 
 instance ToJSON Reference where
-  toJSON (Reference ref) = object [ "$ref" .= ref ]
+  toJSON (Reference ref) = object [ "$ref" .= percentEncodeT ref ]
```

Comment:
This seems to me the wrong place to do this.  The text inside the `Reference` should really be a type that can only be turned into a string by encoding it; the fact that the string is *also* encoded if it appears inside a json object is coincidental.

Reply:
I'm not sure I understand what you are trying to say.

`Reference` references some other type, as you have alluded to.

IMO the Haskell data type should contain the literate string, which is the name of the referenced type. 

So, if the type referenced is `User'`, then the literal string contained in the `Reference` should be `"User'"`. However, swagger requires all URIs (references) to be a proper URI, which requires special characters (such as `'`) to be escaped with percent-encoding. So, percent-encoding really is just the way the `Reference` type is converted to swagger.

And, as far as I understand it, in the scope of this library, `ToJSON` instances on swagger-related datatypes should produce a valid swagger description.

Reply:
I will fix the tests (if that will be necessary), once we come to a common understanding.